rmf_site_editor/src/site/model.rs

```diff
@@ -16,528 +16,649 @@
-) {
-    for (scale, scene) in changed_scales.iter() {
-        if let Some(scene) = scene.entity {
-            if let Ok(mut tf) = transforms.get_mut(scene) {
-                tf.scale = **scale;
+#[derive(Clone, Debug, PartialEq, Eq, Hash, DeliveryLabel)]
+struct SpawnModelLabel(Entity);
+
+/// Component added to models that are being loaded
+#[derive(Component, Deref, DerefMut)]
+pub struct ModelLoadingState(Promise<ModelLoadingResult>);
+
+/// Component added to models that failed loading and containing the reason loading failed.
+#[derive(Component, Deref, DerefMut)]
+pub struct ModelFailedLoading(ModelLoadingError);
+
+/// Polling system that checks the state of promises and prints errors / adds marker components if
+/// models failed loading
+fn handle_model_loading_errors(
+    In(result): In<ModelLoadingResult>,
+    model_scenes: Query<&ModelScene>,
+    mut commands: Commands,
+) -> ModelLoadingResult {
+    let parent = match result {
+        Ok(ref success) => success.request.parent,
+        Err(ref err) => {
+            let parent = err.request.parent;
+            // There was an actual error, cleanup the scene
+            if let Ok(scene) = model_scenes.get(parent) {
+                commands.entity(scene.scene_root).despawn_recursive();
+                commands.entity(parent).remove::<ModelScene>();
             }
+            error!("{err}");
+            commands
+                .entity(parent)
+                .insert(ModelFailedLoading(err.clone()));
+            parent
         }
+    };
+    commands.entity(parent).remove::<ModelLoadingState>();
+    result
+}
+
+/// `SystemParam` used to request for model loading operations
+#[derive(SystemParam)]
+pub struct ModelLoader<'w, 's> {
+    services: Res<'w, ModelLoadingServices>,
+    commands: Commands<'w, 's>,
+}
+
+impl<'w, 's> ModelLoader<'w, 's> {
+    /// Spawn a new model and begin a workflow to load its asset source.
+    /// This is only for brand new models does not support reacting to the load finishing.
+    pub fn spawn_model(&mut self, parent: Entity, model: Model) -> EntityCommands<'w, 's, '_> {
+        self.spawn_model_impulse(parent, model, move |impulse| {
+            impulse.detach();
+        })
+    }
+
+    /// Spawn a new model and begin a workflow to load its asset source.
+    /// Additionally build on the impulse chain of the asset source loading workflow.
+    pub fn spawn_model_impulse(
+        &mut self,
+        parent: Entity,
+        model: Model,
+        impulse: impl FnOnce(Impulse<ModelLoadingResult, ()>),
+    ) -> EntityCommands<'w, 's, '_> {
+        let source = model.source.clone();
+        let id = self.commands.spawn(model).set_parent(parent).id();
+        let loading_impulse = self.commands.request(
+            ModelLoadingRequest::new(id, source),
+            self.services
+                .load_model
+                .clone()
+                .instruct(SpawnModelLabel(id).preempt()),
+        );
```

Comment:
```suggestion
        let loading_impulse = self.update_asset_source_impulse(id, source);
```
We can remove some duplication here so if we ever decide to tweak the model loading command, it will be tweaked for both functions at once.